### DISCUSSION TITLE
AudioConverter.convert conversion settings help

### DISCUSSION THREAD

**Discussion by bjnortier** [sbooth/SFBAudioEngine#293] [Q&A] [answered]
Hi, I'm struggling a bit with all the audio settings and lack of knowledge, but I want to convert audio to WAV 16bit, 16kHz, 1 channel, non-interleaved. I think the code below is on the right track but it throws an error & I don't know how to specify bitrate & non-interleaved. 

```
        let output = OutputSource.makeForData()
        let encoder = try AudioEncoder(outputSource: output, encoderName: .coreAudio)
        encoder.settings = [
            .coreAudioFileTypeID: kAudioFileWAVEType,
            .coreAudioFormatID: kAudioFormatLinearPCM,
            .coreAudioBitsPerChannel: 16
        ]
        try AudioConverter.convert(inputURL, using: encoder)
```

Any help would be appreciated!

**Comment by sbooth** [comment, marked as answer]
To adjust properties like the sample rate you can create an instance of `AudioConverter` using the following constructor:

https://github.com/sbooth/SFBAudioEngine/blob/838d90966067f395929f2df558e6925c8ca24577/Conversion/SFBAudioConverter.h#L115

The proposed format passed to the block can be inspected and a different format returned if desired.

Further customization can also be accomplished by setting properties on the intermediate converter:

https://github.com/sbooth/SFBAudioEngine/blob/838d90966067f395929f2df558e6925c8ca24577/Conversion/SFBAudioConverter.h#L126

For the situation you describe using the Core Audio encoder would look like:

```
let decoder = try AudioDecoder(url: url)

let encoder = try AudioEncoder(url: destURL, encoderName: .coreAudio)
encoder.settings = [
	.coreAudioFileTypeID: kAudioFileWAVEType,
	.coreAudioFormatID: kAudioFormatLinearPCM,
	.coreAudioBitsPerChannel: 16,
	.coreAudioFormatFlags: kAudioFormatFlagIsSignedInteger | kAudioFormatFlagIsPacked
]

let converter = try AudioConverter(decoder: decoder, encoder: encoder) { _ in
	return AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: 16000, channels: 1, interleaved: false)!
}

try converter.convert()
```

You could also use the Libsndfile encoder:

```
let decoder = try AudioDecoder(url: url)

let encoder = try AudioEncoder(url: destURL, encoderName: .libsndfile)
encoder.settings = [
	.libsndfileMajorFormat: LibsndfileMajorFormat.WAV,
	.libsndfileSubtype: LibsndfileSubtype.PCM_16
]

let converter = try AudioConverter(decoder: decoder, encoder: encoder) { _ in
	return AVAudioFormat(commonFormat: .pcmFormatInt16, sampleRate: 16000, channels: 1, interleaved: false)!
}

try converter.convert()
```

**Comment by bjnortier** [comment]
This is amazing. Tested it and it works. Thank you!
